### PR TITLE
Source Facebook Marketing: expand time_increment config, add date_preset config

### PR DESCRIPTION
**Pull request by scottleechua**
## What

Solves Issue #14391:

"When fetching reach data, we oftentimes want monthly... or even lifetime aggregations over campaigns or ads. With the current [Airbyte] implementation of `time_increment`, this is not possible. The Insights API does [allow for such specifications though](https://developers.facebook.com/docs/marketing-api/insights/parameters/v14.0), using the `date_preset` and `time_increment` parameters.

- To fetch lifetime data we would set `date_preset` = `maximum` and `time_increment` = `all_days`
- To fetch monthly data we would set `date_preset` = `maximum` and `time_increment` = `monthly`

The `time_increment` configuration currently only allows for values [1,89] in days and defaults to 1."

## How

- Expand `time_increment` type from positive integer only, to also include the exact strings `all_days` and `monthly`
- Add attribute `date_preset = maximum` to the `AdsInsights` stream.
- Update date interval handling to allow monthly increments.

## 🚨 User Impact 🚨
No breaking changes. User can optionally use `all_days` and `monthly` as `time_increment`.

## Pre-merge Actions
<details><summary><strong>Updating a connector</strong></summary>

All existing unit tests pass:

Screenshot 2023-05-01 at 3 06 06 PM

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
@scottleechua you must add unit test to validate all these changes. Please let me know if you need any assistance to add them.

**Comment by scottleechua** [comment]
Hi @marcosmarxm, would greatly appreciate help on the unit tests!

My thought is to add two more test states containing:
1. `"time_increment": "monthly"`
2. `"time_increment": "all_days"`

to `unit_tests/test_base_insight_streams.py`:

```
    @pytest.mark.parametrize(
        "state",
        [
            {
                AdsInsights.cursor_field: "2010-10-03",
                "slices": [
                    "2010-01-01",
                    "2010-01-02",
                ],
                "time_increment": 1,
            },
            {
                AdsInsights.cursor_field: "2010-10-03",
            },
            {
                "slices": [
                    "2010-01-01",
                    "2010-01-02",
                ]
            },
        ],
    )
```

but it's not clear what the current unit tests are doing.

Also would need help testing this as I don't have a Facebook Marketing sandbox API 🙏 

What do you think?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py [hunk 1]
```diff
@@ -60,7 +60,8 @@ def __init__(
         breakdowns: List[str] = None,
         action_breakdowns: List[str] = None,
         action_breakdowns_allow_empty: bool = False,
-        time_increment: Optional[int] = None,
+        time_increment: Optional[Union[int,str]] = None,
+        date_preset: str = "maximum",
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py:64]: According to the [FB API doc](https://developers.facebook.com/docs/marketing-api/reference/adgroup/insights/) - `This field is ignored if time_range or time_ranges is specified.`, so I guess this wouldn't work as expected

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py [hunk 2]
```diff
@@ -175,11 +177,18 @@ def get_updated_state(self, current_stream_state: MutableMapping[str, Any], late
...
+        if an integer, increment over date range by days.
+        """
         if self._end_date < self._next_cursor_value:
             return
         date_range = self._end_date - self._next_cursor_value
-        yield from date_range.range("days", self.time_increment)
+        if self.time_increment == "monthly":
+            yield from date_range.range("months")
+        else:
+            yield from date_range.range("days", self.time_increment)
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py:191]: very likely you will get an error here if `time_increment == "all_days"`

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py [hunk 3]
```diff
@@ -198,10 +207,15 @@ def _generate_async_jobs(self, params: Mapping) -> Iterator[AsyncJob]:
         """
 
         self._next_cursor_value = self._get_start_date()
+        if self.time_increment == "all_days":
+            pass
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py:211]: what's the sense of this statement?

Comment (davydov-d) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py:211]: did you mean to add `else` afterwards?

**Review by davydov-d — CHANGES_REQUESTED** [review]
@scottleechua thanks for you effort. I've left some comments, also I have a feeling that we're trying to mix things up. The `time_increment` option we used to have in the config is not the equivalent of the `time_increment` in FB API. We used the first one to slice date range, but the second one is rather a `group by` period from my perspective. 
Another request is to cover such changes with a new entry in CAT config and unit tests as well; also current state of the PR needs the conflict to be resolved.

**Comment by bmcgilli** [comment]
Hey @scottleechua! Are you still able to work with @davydov-d on this contribution? Please let us know or we will close it, thanks

**Comment by scottleechua** [comment]
@bmcgilli @davydov-d Sorry, I'm not able to continue work on this! Please go ahead and close the PR. 🙏

**Pull request closed without merging**